src/hybrid_reader.c

```diff
@@ -210,6 +215,30 @@ static void prepareResults(HybridIterator *hr) {
     if (heap_count(hr->topResults) == hr->query.k) {
       break;
     }
+    // Re-evaluate the child num estimated results and the hybrid policy based on the current batch.
+    size_t new_results_cur_batch = heap_count(hr->topResults) - (hr->query.k - n_res_left);
+    // This is the ratio between index_size to child results size as reflected by this batch.
+    float cur_ratio = (float)new_results_cur_batch / n_res_left;
+    // Child estimated number of results as reflected by this batch.
+    size_t cur_child_num_estimated = cur_ratio * VecSimIndex_IndexSize(hr->index);
+    // Conclude the new estimation of the child res num as the average between the old
+    // and new estimation (get the accumulated estimation).
+    child_num_estimated = (child_num_estimated + cur_child_num_estimated) / 2;
+    if (child_num_estimated > child_upper_bound) {
+      child_num_estimated = child_upper_bound;
+    }
+    if (VecSimIndex_PreferAdHocSearch(hr->index, child_num_estimated, hr->query.k, false)) {
+      // Change policy from batches to AD-HOC BF.
+      VecSimBatchIterator_Free(batch_it);
+      hr->searchMode = VECSIM_HYBRID_BATCHES_TO_ADHOC_BF;
```

Comment:
This is taken care of by `initial_check` param, no?

Reply:
right. `initial_check=false` here means that we revisit our hybrid policy (rather than selecting it for the first time)